source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,27 +236,711 @@ namespace rex
+        {
+          auto it = rsl::max_element(gpus.cbegin(), gpus.cend(),
+                                     [](const GpuDescription& lhs, const GpuDescription& rhs)
+                                     {
+                                       const size_t lhs_vram = lhs.dedicated_video_memory.size_in_bytes();
+                                       const size_t rhs_vram = rhs.dedicated_video_memory.size_in_bytes();
+
+                                       return rhs_vram > lhs_vram;
+                                     });
+
+          return it != gpus.cend() ? rsl::distance(gpus.cbegin(), it) : -1;
+        }
+      } // namespace internal
+
+      //-------------------------------------------------------------------------
+      bool flush_command_queue()
+      {
+        // Advance the fence value to mark commands up to this fence point.
+        g_ctx.current_fence++;
+
+        // Add an instruction to the command queue to set a new fence point. Because we
+        // are on the GPU timeline, the new fence point won't be set until the GPU finishes
+        // processing all the commands prior to this Signal().
+        if(FAILED(g_ctx.command_queue->Signal(g_ctx.fence.Get(), g_ctx.current_fence)))
+        {
+          REX_ERROR(LogDirectX, "Failed to signal command queue with fence");
+          return false;
+        }
+
+        if(g_ctx.fence->GetCompletedValue() < g_ctx.current_fence)
+        {
+          HANDLE event_handle = CreateEventEx(nullptr, false, false, EVENT_ALL_ACCESS);
+
+          // Fire event when GPU hits current fence
+          if(FAILED(g_ctx.fence->SetEventOnCompletion(g_ctx.current_fence, event_handle)))
+          {
+            REX_ERROR(LogDirectX, "Failed to set completion event for fence");
+            return false;
+          }
+
+          WaitForSingleObject(event_handle, INFINITE);
+          CloseHandle(event_handle);
+        }
+
+        return true;
+      }
+
       //-------------------------------------------------------------------------
-      bool initialize()
+      bool initialize(const OutputWindowUserData& userData, s32 maxFrameResources, const ResourceSlot& fbColorTargetSlot, const ResourceSlot& bbColorTargetSlot, const ResourceSlot& depthTargetSlot)
       {
-        dxgi::Factory factory = dxgi::Factory::create();
+        // Diagnostic Layer
+#if defined(REX_DEBUG)
+        // Enable extra debuggin and send debug messages to the VC++ output window
+        rex::wrl::com_ptr<ID3D12Debug> debug_controller;
+        if(FAILED(D3D12GetDebugInterface(IID_PPV_ARGS(&debug_controller))))
+        {
+          REX_ERROR(LogDirectX, "Failed to create DX debug controller");
+          return false;
+        }
+
+        debug_controller->EnableDebugLayer();
+#endif
+
+        // Setup frame context
+        if(maxFrameResources <= 0)
+        {
+          REX_ERROR(LogDirectX, "Unable to initialize renderer when no frame resources are allocated.");
```

Comment:
No I am checking against the amount of "FrameResources" I have. A Frame Resource is just a container for a set amount of constant buffers and a fence. This will be extended in the future to containing render target, depth target and other things I can't really thing of now. I think you are confusing this with "Frame Buffers" , which is not what this is. 

A better name might be required if it is confusing to you other people might be confused about this as well.